File: sdk/kotlin/src/test/kotlin/dev/qenlo/QenloTest.kt
package dev.qenlo

import org.junit.jupiter.api.io.TempDir
import java.nio.file.Path
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertTrue

class QenloTest {
    private fun fixture() = listOf(
        Record(9u, 7u, -5, floatArrayOf(1f, 0f, 0f)),
        Record(2u, 7u, 0, floatArrayOf(2f, 0f, 0f)),
        Record(4u, 8u, 10, floatArrayOf(0f, 1f, 0f)),
        Record(6u, 7u, 20, floatArrayOf(0f, 0f, 1f)),
    )

    @Test fun `typed filters ordering and telemetry`() {
        QenloCollection.memory(3).use { db ->
            db.addBatch(fixture())
            val response = db.search(floatArrayOf(1f, 0f, 0f), Filter(7u, -5, 20))
            assertEquals(listOf(2uL, 9uL), response.results.map(SearchResult::id))
            assertEquals("Cpu", response.report.actualBackend)
            assertEquals("Exact", response.report.algorithm)
            assertTrue(response.report.operationId > 0u)
        }
    }

    @Test fun `atomic batches and non reusable ids`() {
        QenloCollection.memory(3).use { db ->
            db.add(fixture()[0])
            assertFailsWith<QenloException> { db.addBatch(listOf(fixture()[1], fixture()[0])) }
            assertEquals(1, db.stats().rows)
            db.delete(9u)
            assertFailsWith<QenloException> { db.add(fixture()[0]) }
        }
    }

    @Test fun `durable reopen`(@TempDir root: Path) {
        val path = root.resolve("vectors.qenlo")
        QenloCollection.create(path, 3).use { db -> db.addBatch(fixture()); db.deleteBatch(listOf(2u, 4u)); db.flush() }
        QenloCollection.open(path, 3).use { db -> assertEquals(2, db.stats().liveRows) }
    }

    @Test fun `portable qn round trip`(@TempDir root: Path) {
        val path = root.resolve("vectors.qn")
        QenloCollection.memory(3).use { db ->
            db.addBatch(fixture()); db.delete(9u); db.exportQn(path)
            assertFailsWith<QenloException> { db.exportQn(path) }
        }
        QenloCollection.importQn(path, 3).use { db ->
            assertEquals(5uL, db.stats().generation)
            assertEquals(3, db.stats().liveRows)
        }
    }

    @Test fun `validation and lifecycle`() {
        val db = QenloCollection.memory(3)
        assertFailsWith<IllegalArgumentException> { db.add(Record(1u, 1u, 0, floatArrayOf(1f))) }
        assertFailsWith<IllegalArgumentException> { db.search(floatArrayOf(1f, 0f, 0f), k = 0) }
        assertFailsWith<QenloException> { db.add(Record(1u, 1u, 0, floatArrayOf(0f, 0f, 0f))) }
        db.close(); db.close()
        assertFailsWith<QenloException> { db.stats() }
    }
}

File: sdk/typescript/test/qenlo.test.ts
import assert from "node:assert/strict";
import { mkdtempSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import test from "node:test";
import { Collection, QenloError, type RecordInput } from "../src/index.js";

const records: readonly RecordInput[] = [
  { id: 9n, userId: 7n, timestamp: -5n, vector: [1, 0, 0] },
  { id: 2n, userId: 7n, timestamp: 0n, vector: [2, 0, 0] },
  { id: 4n, userId: 8n, timestamp: 10n, vector: [0, 1, 0] },
  { id: 6n, userId: 7n, timestamp: 20n, vector: [0, 0, 1] },
];

test("typed filters, deterministic ordering, and telemetry", () => {
  using db = Collection.memory(3);
  db.addBatch(records);
  const response = db.search([1, 0, 0], { userId: 7n, timestampLower: -5n, timestampUpper: 20n });
  assert.deepEqual(response.results.map((hit) => hit.id), [2n, 9n]);
  assert.equal(response.report.actualBackend, "Cpu");
  assert.equal(response.report.algorithm, "Exact");
  assert(response.report.operationId > 0n);
  assert.equal(db.stats().liveRows, 4);
});

test("atomic batches and non-reusable IDs", () => {
  using db = Collection.memory(3);
  db.add(records[0]!);
  assert.throws(() => db.addBatch([records[1]!, records[0]!]), QenloError);
  assert.equal(db.stats().rows, 1);
  db.delete(9n);
  assert.throws(() => db.add(records[0]!), QenloError);
});

test("durable reopen", () => {
  const root = mkdtempSync(join(tmpdir(), "qenlo-ts-"));
  const path = join(root, "vectors.qenlo");
  try {
    {
      using db = Collection.create(path, 3);
      db.addBatch(records);
      db.deleteBatch([2n, 4n]);
      db.flush();
    }
    using db = Collection.open(path, 3);
    assert.equal(db.stats().liveRows, 2);
    assert.deepEqual(db.search([1, 0, 0]).results.map((hit) => hit.id), [9n, 6n]);
  } finally {
    rmSync(root, { recursive: true, force: true });
  }
});

test("validation and lifecycle failures are typed", () => {
  const db = Collection.memory(3);
  assert.throws(() => db.add({ id: 1n, userId: 1n, timestamp: 0n, vector: [1] }), RangeError);
  assert.throws(() => db.search([1, 0, 0], {}, 0), RangeError);
  assert.throws(() => db.add({ id: 1n, userId: 1n, timestamp: 0n, vector: [0, 0, 0] }), QenloError);
  db.close();
  db.close();
  assert.throws(() => db.stats(), QenloError);
});
